std/experimental/color/rgb.d

```diff
@@ -0,0 +1,546 @@
+    Source:     $(PHOBOSSRC std/experimental/color/rgb.d)
+*/
+module std.experimental.color.rgb;
+
+import std.experimental.color;
+import std.experimental.color.colorspace;
+import std.experimental.color.xyz : XYZ, isXYZ;
+import std.experimental.normint;
+
+import std.traits : isInstanceOf, isNumeric, isIntegral, isFloatingPoint, isSomeChar, Unqual;
+import std.typetuple : TypeTuple;
+import std.typecons : tuple;
+
+@safe pure nothrow @nogc:
+
+
+/**
+Detect whether $(D T) is an RGB color.
+*/
+enum isRGB(T) = isInstanceOf!(RGB, T);
+
+///
+unittest
+{
+    static assert(isRGB!(RGB!("bgr", ushort)) == true);
+    static assert(isRGB!LA8 == true);
+    static assert(isRGB!int == false);
+}
+
+
+// DEBATE: which should it be?
+template defaultAlpha(T)
+{
+/+
+    enum defaultAlpha = isFloatingPoint!T ? T(1) : T.max;
++/
+    enum defaultAlpha = T(0);
+}
+
+
+/**
+An RGB color, parameterised with components, component type, and color space specification.
+
+Params: components_ = Components that shall be available. Struct is populated with components in the order specified.
+                      Valid components are:
+                        "r" = red
+                        "g" = green
+                        "b" = blue
+                        "a" = alpha
+                        "l" = luminance
+                        "x" = placeholder/padding (no significant value)
+        ComponentType_ = Type for the color channels. May be a basic integer or floating point type.
+        linear_ = Color is stored with linear luminance.
+        colorSpace_ = Color will be within the specified color space.
+*/
+struct RGB(string components_, ComponentType_, bool linear_ = false, RGBColorSpace colorSpace_ = RGBColorSpace.sRGB)
+    if(isNumeric!ComponentType_)
+{
+@safe pure nothrow @nogc:
+
+    // RGB colors may only contain components 'rgb', or 'l' (luminance)
+    // They may also optionally contain an 'a' (alpha) component, and 'x' (unused) components
+    static assert(allIn!("rgblax", components), "Invalid Color component '"d ~ notIn!("rgblax", components) ~ "'. RGB colors may only contain components: r, g, b, l, a, x"d);
+    static assert(anyIn!("rgbal", components), "RGB colors must contain at least one component of r, g, b, l, a.");
+    static assert(!canFind!(components, 'l') || !anyIn!("rgb", components), "RGB colors may not contain rgb AND luminance components together.");
+
+    static if(isFloatingPoint!ComponentType_)
+    {
+        /** Type of the color components. */
+        alias ComponentType = ComponentType_;
+    }
+    else
+    {
+        /** Type of the color components. */
+        alias ComponentType = NormalizedInt!ComponentType_;
+    }
+
+    /** The color components that were specified. */
+    enum string components = components_;
+    /** The color space specified. */
+    enum RGBColorSpace colorSpace = colorSpace_;
+    /** If the color is stored linearly (without gamma applied). */
+    enum bool linear = linear_;
```

Comment:
I think I've figured out a way to avoid the underscore suffix in the template parameters:

```d
// For template values
template extractTemplateValue(T, size_t index)
{
    static if (is(T : U!(TL), alias U, TL...))
        enum extractTemplateValue = TL[index];
    else
        static assert(false, "some useful error message");
}

// For template types
template extractTemplateParameter(T, size_t index)
{
    static if (is(T : U!(TL), alias U, TL...))
        alias extractTemplateParameter = TL[index];
    else
        static assert(false, "some useful error message");
}

struct RGB(ComponentType, bool linear = false)
{
    alias ComponentType = extractTemplateParameter!(typeof(this), 0);
    enum linear = extractTemplateValue!(typeof(this), 1);
}
```